sweep/sweeper.go

```diff
@@ -1,40 +1,701 @@
+	// which the fee calculation will be based.
+	SweepTxConfTarget uint32
+
+	// MaxInputsPerTx specifies the default maximum number of inputs allowed
+	// in a single sweep tx. If more need to be swept, multiple txes are
+	// created and published.
+	MaxInputsPerTx int
+
+	// MaxSweepAttempts specifies the maximum number of times an input is
+	// included in a publish attempt before giving up and returning an error
+	// to the caller.
+	MaxSweepAttempts int
+
+	// NextAttemptDeltaFunc returns given the number of already attempted
+	// sweeps, how many blocks to wait before retrying to sweep.
+	NextAttemptDeltaFunc func(int) int32
 }
 
-// New returns a new UtxoSweeper instance.
+// Result is the struct that is pushed through the result channel. Callers
+// can use this to be informed of the final sweep result. In case of a remote
+// spend, Err will be ErrRemoteSpend.
+type Result struct {
+	// Err is the final result of the sweep. It is nil when the input is
+	// swept successfully by us. ErrRemoteSpend is returned when another
+	// party took the input.
+	Err error
+
+	// Tx is the transaction that spent the input.
+	Tx *wire.MsgTx
+}
+
+// sweepInputMessage structs are used in the internal channel between the
+// SweepInput call and the sweeper main loop.
+type sweepInputMessage struct {
+	input      Input
+	resultChan chan Result
+}
+
+// New returns a new Sweeper instance.
 func New(cfg *UtxoSweeperConfig) *UtxoSweeper {
+
 	return &UtxoSweeper{
-		cfg: cfg,
+		cfg:           cfg,
+		newInputs:     make(chan *sweepInputMessage),
+		spendChan:     make(chan *chainntnfs.SpendDetail),
+		quit:          make(chan struct{}),
+		pendingInputs: make(map[wire.OutPoint]*pendingInput),
 	}
 }
 
+// Start starts the process of constructing and publish sweep txes.
+func (s *UtxoSweeper) Start() error {
+	if !atomic.CompareAndSwapUint32(&s.started, 0, 1) {
+		return nil
+	}
+
+	log.Tracef("Sweeper starting")
+
+	// Retrieve last published tx from database.
+	lastTx, err := s.cfg.Store.GetLastPublishedTx()
+	if err != nil {
+		return err
+	}
+
+	// Republish in case the previous call crashed lnd. We don't care about
+	// the return value, because inputs will be reoffered and retried
+	// anyway. The only reason we republish here is to prevent the corner
+	// case where lnd goes into a restart loop because of a crashing publish
+	// tx where we keep deriving new output script. By publishing and
+	// possibly crashing already now, we haven't derived a new output script
+	// yet.
+	if lastTx != nil {
+		log.Debugf("Publishing last tx %v", lastTx.TxHash())
+
+		// Error can be ignored. Because we are starting up, there are
+		// no pending inputs to update based on the publish result.
+		err := s.cfg.PublishTransaction(lastTx)
+		if err != nil {
+			log.Errorf("Last tx publish: %v", err)
+		}
+	}
+
+	// Retrieve relay fee for dust limit calculation. Assume that this will
+	// not change from here on.
+	s.relayFeePerKW = s.cfg.Estimator.RelayFeePerKW()
```

Comment:
since this is only used for calculating the dust limit(?), can we instead store the dust limit directly here?

Reply:
I wanted to have the dust limit calculation inside txgenerator to logically group the functionality together. `generateInputPartitionings` takes two fees and uses that as the basis to calculate the partitionings.